source/text/Font.h

```diff
@@ -44,54 +69,165 @@ class Font {
-	int Width(const std::string &str, char after = ' ') const;
-	// Get the width of the text while accounting for the desired layout and truncation strategy.
-	int FormattedWidth(const DisplayText &text, char after = ' ') const;
-	
-	int Height() const noexcept;
+	// Determine the string's width and height, without considering formatting.
+	int Width(const std::string &str) const;
+	int Height() const;
+	// Get the width and height of the text while accounting for the desired layout and truncation strategy.
+	int FormattedWidth(const DisplayText &text) const;
+	int FormattedHeight(const DisplayText &text) const;
 	
-	int Space() const noexcept;
+	// Get the line height and paragraph break.
+	int LineHeight(const Layout &layout = {}) const;
+	int ParagraphBreak(const Layout &layout = {}) const;
 	
 	static void ShowUnderlines(bool show) noexcept;
 	
+	// Escape markup characters if it causes some errors.
+	static std::string EscapeMarkupHasError(const std::string &str);
 	
 private:
-	static int Glyph(char c, bool isAfterSpace) noexcept;
-	void LoadTexture(ImageBuffer &image);
-	void CalculateAdvances(ImageBuffer &image);
-	void SetUpShader(float glyphW, float glyphH);
+	// Text rendered as a sprite.
+	struct RenderedText {
+		// Texture with rendered text.
+		GLuint texture;
+		int width;
+		int height;
+		// Offset from the floored origin to the center of the sprite.
+		Point center;
+	};
 	
-	int WidthRawString(const char *str, char after = ' ') const noexcept;
+	// A key mapping the text and layout parameters, underline status to RenderedText.
+	struct CacheKey {
+		DisplayText text;
+		bool showUnderline;
+		
+		CacheKey(const DisplayText &t, bool underline) noexcept;
+		bool operator==(const CacheKey &rhs) const noexcept;
+	};
 	
-	std::string TruncateText(const DisplayText &text, int &width) const;
-	std::string TruncateBack(const std::string &str, int &width) const;
-	std::string TruncateFront(const std::string &str, int &width) const;
-	std::string TruncateMiddle(const std::string &str, int &width) const;
+	// Hash function of CacheKey.
+	struct CacheKeyHash {
+		typedef CacheKey argument_type;
+		typedef std::size_t result_type;
+		result_type operator() (argument_type const &s) const noexcept;
+	};
+	
+	// Function to recycle it for RenderedText.
+	class AtRecycleForRenderedText {
+	public:
+		void operator()(RenderedText &v) const;
+	};
+	
+	
+private:
+	void UpdateSurfaceSize() const;
+	void UpdateFont() const;
+	
+	static std::string ReplaceCharacters(const std::string &str);
+	static std::string RemoveAccelerator(const std::string &str);
+	
+	void DrawCommon(const DisplayText &text, double x, double y, const Color &color, bool alignToDot) const;
+	const RenderedText &Render(const DisplayText &text) const;
+	void SetUpShader();
+	
+	int ViewWidth(const DisplayText &text) const;
+	
+	// Convert Viewport to/from Text coordinates.
+	double ViewFromTextX(double x) const;
+	double ViewFromTextY(double y) const;
```

Comment:
Though the generally accepted variable name for simple math is e.g. x, y, or z, here we have same-axis conversions, so we should specify a bit more clearly which "world" the function parameter is from, e.g. `___ViewportX(double textX)` and `___TextX(double viewportX)` where "___" is e.g. "To", "Get", "As", or similar.